### PR TITLE
PR

### PR DESCRIPTION
**Pull request by vmarsi**

### CODE REVIEW

File: src/flood_wave_detector.py [hunk 1]
```diff
@@ -71,10 +72,48 @@ def run(self) -> None:
...
+
+            # Set original values
+            self.gauges = gauges_copy
+            self.start_date = cut_dates[0]
+            self.end_date = cut_dates[-1]
+
         GraphBuilder().build_graph(folder_name=self.folder_name)
 
+    @staticmethod
+    def add_elem(list0: list, date: str):
```
Comment (zsvizi) [inline src/flood_wave_detector.py]: You can remove this method and replace its call with the syntax: `list0 + [date] if date not in list0 else list0`

File: src/flood_wave_detector.py [hunk 2]
```diff
@@ -130,8 +177,9 @@ def find_edges(self) -> None:
             does_actual_json_exist = os.path.exists(os.path.join(PROJECT_PATH, self.folder_name, 'find_edges',
                                                     f'{current_gauge}_{next_gauge}.json'))
 
-            if does_actual_json_exist and does_big_json_exist:
-                continue
+            # I had to comment this part out because from now on updating existing files is crucial.
```
Comment (zsvizi) [inline src/flood_wave_detector.py]: Remove this part then.

File: src/dataloader.py [hunk 3]
```diff
@@ -21,6 +21,10 @@ def download_data(self):
             url = 'https://drive.google.com/uc?id=1jZP1bfGTjoz-f-VzHPmzH6YU6ZaO_JQa'
             output = os.path.join(PROJECT_PATH, 'data', self.dataset_name + ".csv")
             gdown.download(url, output, quiet=False)
+        if not os.path.exists(os.path.join(PROJECT_PATH, 'data', 'existing_stations' + ".json")):
```
Comment (zsvizi) [inline src/dataloader.py]: In theory, `gdown` can handle link to shared folders - could you take a look at it? If it does, then collapse all downloads into one block.

File: src/flood_wave_detector.py [hunk 4]
```diff
@@ -71,10 +72,48 @@ def run(self) -> None:
         Executes the steps needed to find all the flood waves.
         """
         self.mkdirs()
-        self.find_vertices()
-        self.find_edges()
+
+        stations_life_intervals = JsonHelper.read(filepath=os.path.join(PROJECT_PATH,
+                                                                        'data', 'existing_stations.json'))
+        all_dates = [self.start_date, self.end_date]
```
Comment (zsvizi) [inline src/flood_wave_detector.py]: I would put code in LOC78-85 to a static method in `FloodWaveHandler`.

**Review by zsvizi — CHANGES_REQUESTED** [review]

**Comment by zsvizi** [comment]
The `Filtering` class seems to be a static class (all methods are static), since instead of calling `self.is_gauge_in_comp` you can write `Filtering.is_gauge_in_comp`, then `self` arguments will not be needed.

**Comment by zsvizi** [comment]
Since we have a couple of filtering methods in `Filtering` and everything is static, now it makes sense to create a `FilteringHandler` class and put there all the utility-like methods from `Filtering` (e.g. `remove_nodes`).

**Comment by zsvizi** [comment]
After all changes, we will have a couple of functionalities in the code:
- data loading and handling
- flood wave detection
- filtering
My request is to create modules in `src` with names `data,` `flood_wave_core`, and `filtering` for these functionalities. Additionally, we can have a `utils` module for `json_helper.py` and `measure_time.py`. Finally, `analysis.py` can stay in `src`. Use the *Refactor* option of PyCharm to move the files into different folders, thus you do not need to calibrate the imports by hand. Take care about the notebook too, PyCharm does not do refactor in the notebooks.

File: src/flood_wave_detector.py [hunk 5]
```diff
@@ -71,8 +72,36 @@ def run(self) -> None:
...
+        stations_life_intervals = JsonHelper.read(filepath=os.path.join(PROJECT_PATH,
+                                                                        'data', 'existing_stations.json'))
+
+        cut_dates = FloodWaveHandler.get_dates_in_between(start_date=self.start_date,
+                                                          end_date=self.end_date,
+                                                          intervals=stations_life_intervals,
+                                                          gauges=self.gauges)
+
+        for i in range(len(cut_dates) - 1):
+            gauges_copy = self.gauges
```
Comment (zsvizi) [inline src/flood_wave_detector.py]: Are you sure that this stores a COPY of `self.gauges` and not a REFERENCE to it? I would use `.copy()` or `copy.deepcopy(self.gauges)` from package `copy`.

**Review by zsvizi — CHANGES_REQUESTED** [review]

File: src/flood_wave_detector.py (hunk 5 above)
Comment (vmarsi) [reply src/flood_wave_detector.py]: I noticed that this part should be outside of the loop anyway. But I did the deepcopy thing too.

**Review by zsvizi — APPROVED** [review]

**Pull request merged**